tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,610 @@
+             (element_type == other.element_type);
+    }
+  };
+
+  struct MergingShapeHash {
+    std::size_t operator()(const MergingShape& dotShape) const {
+      std::size_t hash = dotShape.m_dim.hash();
+      hash = llvm::hash_combine(hash, dotShape.n_dim.hash());
+      hash = llvm::hash_combine(hash, dotShape.contracting_dim.hash());
+      for (const auto& d : dotShape.batching_dims) {
+        hash = llvm::hash_combine(hash, d.hash());
+      }
+      const auto& dimension_numbers = dotShape.dimension_numbers;
+      const auto& lhs_batch_dims = dimension_numbers.getLhsBatchingDimensions();
+      const auto& rhs_batch_dims = dimension_numbers.getRhsBatchingDimensions();
+      const auto& lhs_contracting_dims =
+          dimension_numbers.getLhsContractingDimensions();
+      const auto& rhs_contracting_dims =
+          dimension_numbers.getRhsContractingDimensions();
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(lhs_batch_dims.begin(),
+                                                         lhs_batch_dims.end()));
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(rhs_batch_dims.begin(),
+                                                         rhs_batch_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(lhs_contracting_dims.begin(),
+                                         lhs_contracting_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(rhs_contracting_dims.begin(),
+                                         rhs_contracting_dims.end()));
+      hash = llvm::hash_combine(hash, mlir::hash_value(dotShape.element_type));
+      return hash;
+    }
+  };
+
+  using MergingShapeEqualMap =
+      std::unordered_map<MergingShape, SmallVector<mhlo::DotGeneralOp>,
+                         MergingShapeHash>;
+
+ private:
+  bool buildMergingShapeMap(Block* block, ShapeAnalysis& analysis,
+                            MergingShapeEqualMap& equal_merge_shape_map);
+  bool applyMerging(DotCluster& cluster);
+  Value expandDim0(OpBuilder& builder, Location& loc, Value value);
+
+ private:
+  FuncOp func_;
+};
+
+bool DotBatchMergeConverter::run() {
+  ShapeAnalysis analysis(func_);
+  if (failed(analysis.run())) {
+    LLVM_DEBUG(llvm::dbgs() << "ShapeAnalysis failes for dot merge.\n");
+    return false;
+  }
+
+  SmallVector<Block*> blocks;
+  func_.walk([&](Block* block) { blocks.push_back(block); });
+
+  for (Block* block : blocks) {
+    // A map to help to cluster dots with same shape and dim-numbers together.
+    MergingShapeEqualMap equal_merge_shape_map;
+    if (!buildMergingShapeMap(block, analysis, equal_merge_shape_map)) {
+      continue;
+    }
+    // Find merging clusters.
+    SmallVector<DotCluster> merging_clusters;
+    BuildDotClusters<MergingShapeEqualMap>(block, equal_merge_shape_map,
```

Comment:
is it necessary for BuildDotClusters() to be a template function?

Reply:
> is it necessary for BuildDotClusters() to be a template function?

This function is designed to be reused by same-operand-dot-merging. My same-operand-dot-merging implementation is here: https://github.com/JamesTheZ/BladeDISC/commit/73442372b5204f318787917371d2d0c0ebd45a8a#diff-37107805397b8e79aed36a0b338da94b5ef55757b91710e73852f579bda03f2fR148